### ISSUE TITLE
Insuring interoperability of pint with other (non-unit) array-like types

### ISSUE THREAD

**Issue by jthielen** [hgrecco/pint#845]
When working with the new `__array_function__` implementation in #764, seeing what it takes to get pint `Quantity`'s working inside xarray (https://github.com/pydata/xarray/issues/525), and trying to wrap my head around [NEP 18](https://www.numpy.org/neps/nep-0018-array-function-protocol.html), I realized that (to the best of what I could find) pint's expected behavior with other ndarray-like types is not well-documented or tested for. And so, I hope that this issue could be a discussion on clarifying those expected behaviors and determining how to document and test them.

To start that discussion, I think it would make sense to split other array-like types into two categories, those that should wrap pint, and those that should be wrapped by pint, or in other words, determine where pint should fall in the "[type casting hierarchy](https://www.numpy.org/neps/nep-0013-ufunc-overrides.html#type-casting-hierarchy)". Based on https://github.com/pydata/xarray/issues/525#issuecomment-482651181 and other comments I've seen, a split like the following seems to be desired:

**Wrap pint Quantity**

- xarray DataArray

**Wrapped by pint Quantity**

- dask array
- cupy array
- sparse array
- numpy masked array (and of course ndarray)

(Pandas seems to have special handling through pint-pandas, and hopefully others that I'm missing can be added to either list as appropriate)

Does this make sense as a division, and should a list like this be documented anywhere?

Also, once #764 is in place, it seems like there will be a need for explicit identification of types in each category...types that should wrap pint are used in https://github.com/hgrecco/pint/blob/4dcbe78766269f985d1898601d8f5384a1511f2c/pint/quantity.py#L70, and types that should be wrapped by pint would be used with implementing the functionality mentioned in https://github.com/hgrecco/pint/pull/764/#issuecomment-523272038. However, what should be done about array-like types that are unknown to pint, and how can consistent behavior between different `Quantity`-related operations be assured?

Finally, should pint implement tests for all the types it expects to be able to wrap? I got the impression from https://github.com/pydata/xarray/issues/525 and https://github.com/pydata/xarray/pull/2956 that tests fit best with the type higher in the type casting hierarchy (the one doing the wrapping).

Please do correct me on anything I am mistaken about with the above rambling :smile:.

xref https://github.com/pydata/xarray/issues/525, https://github.com/pydata/xarray/pull/2956, https://github.com/hgrecco/pint/pull/764/, https://github.com/hgrecco/pint/issues/633, https://github.com/Unidata/MetPy/issues/996

EDIT: Discussion about arbitrarily nested metadata with NEP 18 implementers is ongoing at https://github.com/dask/dask/issues/5329, which will be good to keep in mind here too.

**Comment by crusaderky** [comment]
@shoyer I changed the style of the arrows to dotted where there is only planned or untested support, and added arrows pint -> sparse and pint-> cupy

https://gist.github.com/crusaderky/1b00fd68ae1ce79109b6af546c88bd4a

**Comment by pentschev** [comment]
Those are nice graphs @shoyer, thanks for doing that.

> The general principle here is that if library X can wrap Y arrays, then it should also handle their interaction. This keeps the logic about how different array library interact in one place, as much as possible.

I'm wondering if this would be a guideline that would be worth to add to the NEP-18? I feel it's quite a good idea to make it more visible, even if elsewhere.

**Comment by jthielen** [comment]
Some progress towards this goal (at least within Pint) has been made with #905, however, I think the following are the high-level tasks remaining:

- [ ] Type checks on Quantity creation to ensure type casting hierarchy is maintained (see #479)
- [ ] Rethink `force_ndarray` array behavior in `_to_magnitude` and elsewhere to handle casting to ndarray and casting to duck arrays ([NEP-30](https://numpy.org/neps/nep-0030-duck-array-protocol.html) may be relevant here)
- [ ] Determine if the type casting hierarchy assumption used in #905 is valid (Pint has a limited, pre-defined list of types that are above it in the hierarchy, and it will wrap all other array-like types)
  - Should the list of upcast types be made extensible?
- [ ] Resolve #924
- [ ] Add tests with upcast types to ensure Pint defers (and commutes) properly
  - [ ] Pandas
  - [ ] xarray
- [ ] Add tests with downcast types to ensure Pint wraps (and commutes with) them properly
  - [ ] Sparse
  - [ ] Dask (simple wrapping, further integration can be left to https://github.com/hgrecco/pint/issues/883)
  - [ ] CuPy (is special CI needed for GPU tests? If so, this can probably be deferred)
  - [ ] Masked arrays (see #37, #633, #688)
- [ ] Add documentation on array type compatibility (as discussed here and in #878)

**Comment by hgrecco** [comment]
Thanks @jthielen for summarizing this. I think is important to have input from people using heavily those packages to make it not only correct but also that it feels right. We can start by collecting some tests. 

Another question that I have: should we have related packages (like pint-pandas) as a way to speed up development?

> Some progress towards this goal (at least within Pint) has been made with #905,

A large progress has been made in #905, don't underestimate your work! 😄

**Comment by jthielen** [comment]
In light of #955, I hope to be working on most of the areas listed above in the next couple days. However, the tests with downcast types are likely to not all be ready for 0.10 because

- Dask is held up due to https://github.com/dask/dask/issues/4583
- I don't want to worry about GPU tests to test CuPy

However, I think testing Sparse and masked arrays as downcast types should be sufficient for now.
